docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,239 @@
+
+In reality, if node `C` ha received `tx_1` from multiple senders (`X`,`Y`, `B`), it will pick the first one that has sent it and disable that route. 
+
+When a connection to a peer is lost a node sends a `Reset` message to its peers. It
+signals to the peers that they should delete all routing information related to this node. This assures that, in case the network topology changes, or a peer becomes unavailable, the node is able to discover transactions via a previously disabled route. 
+
+### Redundancy control
+
+In an ideal setting, receiving a transaction more than once is not needed. But in a Byzantine setting, allowing for only one route for a transaction can introduce an attack vector. 
+
+Operators can thus configure a desired redundancy and the gossip protocol will adjust route disabling based on this setting. Redundancy is defined as the ratio between duplicate and unique transactions. Operators can also define how much they tolerate the redundancy to deviate from the desired level.
+
+By default, the redundancy is set to 1 and impacts the gossiping of transactions as follows:
+
+
+```
+  redundancy := duplicateTxs / firstTimeTxs
+  if redundancy < r.redundancyLowerBound:
+    peer.send(Reset)
+  if redundancy > redundancyUpperBound
+    enableRouteDisabling() 
+```
+
+Redundancy is adjusted whenever a new transaction is added into the mempool.
+(TODO - revisit whether this should also be done on duplicates.)
+
+
+### Impacted areas of code
+
+#### Mempool reactor
+
+
+The bulk of the changes is constrained to the mempool reactor. 
+We introduce a new communication channel, called `Mempool Control Channel`. The channel
+is used to transmit the messages needed for the implementation of the protocol. 
+
+Additionally, the mempool reactor is extended with a 
+`gossipRouter` and a `redundancyControl` struct to keep track of the redundancy
+of transactions in the system. 
+Each node is keeps track of the routes it disabled between its peers. 
+
+```
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating transactions. Source
+	// and target are node IDs.
+	disabledRoutes map[nodekey.ID]p2pIDSet
+}
+``` 
+
+
+```
+type redundancyControl struct {
+	txsPerAdjustment int64
+
+	// Pre-computed upper and lower bounds of accepted redundancy.
+	lowerBound float64
+	upperBound float64
+
+	mtx          cmtsync.RWMutex
+	firstTimeTxs int64 // number of transactions received for the first time
+	duplicates   int64 // number of duplicate transactions
+
+	// If true, do not send HaveTx messages.
+	blockHaveTx atomic.Bool
+}
+
+```
+
+
+
+#### New p2p messages
+
+The protocol introduces two new p2p messages whose protobuf definition is given below:
+
+```
+type HaveTx struct {
+	TxKey []byte 
+}
+```
+
+```
+type Reset struct {
+
+}
+```
+
+#### Monitoring
+
+This ADR introduces a set of metrics which can be used to observe the parameters of the protocol. 
+
+- `HaveTxMsgsReceived`
```

Comment:
I don't think we need all these metrics in the final version. I would like to see the redundancy level, but instead of `Redundancy` I would replace it with two new metrics `FirstTimeTxs` and `DuplicateTxs` from where we can compute redundancy. In fact we already have the number of duplicates in `AlreadyReceivedTxs`.

Reply:
I removed the HaveTx and Reset msgs from the metrics but I left redundancy. I think it's still beneficial to have it pre-computed because we reset these values (`FirstTimeTxs` and `DuplicateTxs`) periodically and the operator might not adjust the computation correctly. Not sure if it would be a problem but it's safer if we just compute it.

Reply:
I forgot that `DuplicateTxs` is not the same as `AlreadyReceivedTxs` because the first one is reset on every adjustment. Anyway, we would still need to consider `FirstTimeTxs` and `DuplicateTxs` instead of `Redudancy`.
In many testnets I saw that dividing the `DuplicateTxs` metric by the `FirstTimeTxs` metric is more precise than the `Redundancy` metric.

Reply:
In the end we decided to leave only `Redundancy`. `FirstTimeTxs` and `DuplicateTxs` were useful for the initial experiments but will not be needed for nodes in production.